### PR TITLE
Create Generic Entry class and UniqueEntryList

### PR DESCRIPTION
**Pull request by Graphcalibur**
Companies, Events, and Persons all have some common attributes and methods (e.g. name, tags, isSameEntry()).

Creating a generic Entry class and converting UniquePersonList into UniqueEntryList that can hold any child class of Entry allows Companies and Events to be added into the existing infrastructure while reusing as much of the existing code as possible.

Let's
* create a Entry class that Person inherits from and contains the aforementioned common attributes and methods
* convert UniquePersonList into a UniqueEntryList that can hold any child class of Entry.
* convert the Person-specific exceptions into general Entry exceptions

### CODE REVIEW

File: src/main/java/seedu/address/model/entry/UniqueEntryList.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.entry.exceptions.DuplicateEntryException;
+import seedu.address.model.entry.exceptions.EntryNotFoundException;
+
+/**
+ * A list of persons that enforces uniqueness between its elements and does not allow nulls.
```
Comment (albertsutz) [inline src/main/java/seedu/address/model/entry/UniqueEntryList.java:15]: A List of person should be a list of Entry

File: src/main/java/seedu/address/model/entry/UniqueEntryList.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+     * Returns true if the list contains an equivalent person as the given argument.
+     */
+    public boolean contains(T toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameEntry);
+    }
+
+    /**
+     * Adds a person to the list.
+     * The person must not already exist in the list.
```
Comment (albertsutz) [inline src/main/java/seedu/address/model/entry/UniqueEntryList.java:41]: Should change person to entry

File: src/main/java/seedu/address/model/entry/exceptions/DuplicateEntryException.java [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
 
 /**
  * Signals that the operation will result in duplicate Persons (Persons are considered duplicates if they have the same
  * identity).
  */
-public class DuplicatePersonException extends RuntimeException {
-    public DuplicatePersonException() {
+public class DuplicateEntryException extends RuntimeException {
+    public DuplicateEntryException() {
         super("Operation would result in duplicate persons");
```
Comment (albertsutz) [inline src/main/java/seedu/address/model/entry/exceptions/DuplicateEntryException.java:9]: error message should be duplicate entry. the Javadocs should also be changed.

**Review by albertsutz — CHANGES_REQUESTED** [review]
I feel that there can be some slight changes 

1. We will have 3 big children of entries: Event, Person, and Company, therefore

- I think we should package each accordingly
- e.g. Person.java should be in seedu.address.model.entry.person
- This also makes it easier to organize the attributes that is unique to a Person, which is the phone, email, and address. (Phone.java, Email.java. and Address.java should also be in the person folder)

I think that following the above recommendation can help us organize all three types of entries better.

**Comment by Graphcalibur** [comment]
I think that's a good way of organizing the code but I also think it's something that should be done once we start implementing Companies/Events in a later PR.

I've also fixed the various instances of comments and messages still referring to "Persons" instead of "Entries".

**Review by mradhyaf — COMMENTED** [review]
Looks good I think.

**Review by albertsutz — APPROVED** [review]

**Pull request merged**